Add RLScenario for Avalanche RL

NickLucche opened the pull request:
This PR adds a first implementation of the RLScenario we use in [Avalanche RL](https://github.com/ContinualAI/avalanche-rl). Essentially it adds support for a scenario wrapping streams of `gym.Env` environments.

Due to changes to the old "Strategy" model, I still haven't managed to test this PR within avalanche-rl, as the integration would require a few changes in order to match the new Templates-based structure.
Nonetheless, I would appreciate any feedback on this first implementation and whether it adheres to the new benchmark guidelines.

Also, I would like to raise a few issues regarding typing, as with the current implementation we're not able to implicitly infer the type of the experience (e.g. RLExperience) which is returned when iterating a stream (e.g. https://github.com/NickLucche/avalanche/blob/refactoring-avalanche-rl/tests/benchmarks/scenarios/test_rl_scenario.py#L21).
Perhaps we could borrow some of the Generics definitions from the old implementation of scenarios, or define a few subclasses just for better type hinting in `rl_scenario.py`.

AntonioCarta commented on avalanche/benchmarks/scenarios/rl_scenario.py:
```diff
@@ -9,14 +9,108 @@
+
+class RLExperience(CLExperience):
+
+    def __init__(self, env: Union[Env, Callable[[], Env]], n_envs: int = 1, task_label: int = None, current_experience: int = None, origin_stream=None):
```
can you add some docstrings for documentation?

AntonioCarta commented on avalanche/benchmarks/scenarios/rl_scenario.py:
```diff
@@ -9,14 +9,108 @@
+
+class RLExperience(CLExperience):
+
+    def __init__(self, env: Union[Env, Callable[[], Env]], n_envs: int = 1, task_label: int = None, current_experience: int = None, origin_stream=None):
```
Callable[[], Env] is needed to create environments lazily? Isn't it better to simplify this definition and make the entire stream lazy?

NickLucche replied:
We also need to support environments which handle the change of task internally (e.g. habitat) and whose instance must persist across experiences; I'd rather leave it as is for now

AntonioCarta replied:
makes sense, let's keep it as is.

AntonioCarta commented on avalanche/benchmarks/scenarios/rl_scenario.py:
```diff
@@ -9,14 +9,108 @@
+        super().__init__(current_experience, origin_stream)
+        self.env = env
+        self.n_envs = n_envs
+        # task label to be (optionally) used for training purposes
```
I think we can drop the requirement of having mandatory task labels (eventually set to `None`).

Also, in a multi-task scenario you should also set `use_in_eval=True`

AntonioCarta replied:
This is still a problem, it should be `self.task_label = ExperienceAttribute(task_label, use_in_train=True, use_in_eval=True)`. If `task_labels=None` is you can also decide to not having the attribute or having  `self.task_label = ExperienceAttribute(task_label)`

AntonioCarta commented on avalanche/benchmarks/scenarios/rl_scenario.py:
```diff
@@ -9,14 +9,108 @@
+class RLScenario(CLScenario):
+
+    def __init__(self, envs: List[Env],
+                 n_experiences: int,
```
Do we need the `n_experiences` arguments? Since the constructor takes a list of environments I don't see a need for it. The user can manipulate the list easily to modify the stream before calling the constructor.

AntonioCarta commented on avalanche/benchmarks/scenarios/rl_scenario.py:
```diff
@@ -9,14 +9,108 @@
+                    tr_task_labels[i % len(tr_task_labels)])
+
+        # move to template/strategy?
+        if seed is not None:
```
Setting the seed here is dangerous. Can we pass the seed where it's needed?

AntonioCarta commented on tests/benchmarks/scenarios/test_rl_scenario.py:
```diff
@@ -0,0 +1,26 @@
+from avalanche.benchmarks.scenarios.rl_scenario import RLScenario, RLExperience
+import pytest
```
we don't use pytest

AntonioCarta commented on tests/benchmarks/scenarios/test_rl_scenario.py:
```diff
@@ -0,0 +1,26 @@
+    skip = True
+
+
+@pytest.mark.skipif(skip, reason="Need gym to run these tests")
```
you can use unittest here:
```
    @unittest.skipIf(...)
```

AntonioCarta commented:
> Due to changes to the old "Strategy" model, I still haven't managed to test this PR within avalanche-rl, as the integration would require a few changes in order to match the new Templates-based structure.

Of course, this would be the next step to figure out.

> Perhaps we could borrow some of the Generics definitions from the old implementation of scenarios

I agree that we need to fix the typing. Also, right now mypy is basically useless because it throws a ton of errors.

AntonioCarta commented:
apart from the task labels, everything looks good.

NickLucche commented:
I've added some code to handle the detection of task labels which should cover the removal of the argument `n_experiences` and added a test showing its behavior
